apollo-router-core/src/query_planner/mod.rs

```diff
@@ -253,66 +270,129 @@ impl PlanNode {
-    operation: String,
-}
+    /// A fetch node.
+    #[derive(Debug, PartialEq, Deserialize)]
+    #[serde(rename_all = "camelCase")]
+    pub(crate) struct FetchNode {
+        /// The name of the service or subgraph that the fetch is querying.
+        service_name: String,
 
-impl FetchNode {
-    async fn fetch_node<'a>(
-        &'a self,
-        response: Arc<Mutex<Response>>,
-        current_dir: &'a Path,
-        request: Arc<Request>,
-        service_registry: Arc<dyn ServiceRegistry>,
-        schema: Arc<Schema>,
-    ) -> Result<(), FetchError> {
-        let FetchNode {
-            variable_usages,
-            requires,
-            operation,
-            service_name,
-        } = self;
-
-        let query_span = tracing::info_span!("subfetch", service = service_name.as_str());
-
-        if let Some(requires) = requires {
-            // We already checked that the service exists during planning
-            let fetcher = service_registry.get(service_name).unwrap();
-
-            let mut variables = Object::with_capacity(1 + variable_usages.len());
-            variables.extend(variable_usages.iter().filter_map(|key| {
-                request.variables.as_ref().map(|v| {
-                    v.get(key)
-                        .map(|value| (key.clone(), value.clone()))
-                        .unwrap_or_default()
-                })
-            }));
-
-            {
-                let response = response.lock().await;
-                tracing::trace!(
-                    "Creating representations at path '{}' for selections={:?} using data={}",
-                    current_dir,
-                    requires,
-                    serde_json::to_string(&response.data).unwrap(),
+        /// The data that is required for the subgraph fetch.
+        #[serde(skip_serializing_if = "Option::is_none")]
+        requires: Option<Vec<Selection>>,
+
+        /// The variables that are used for the subgraph fetch.
+        variable_usages: Vec<String>,
+
+        /// The GraphQL subquery that is used for the fetch.
+        operation: String,
+    }
+
+    struct Variables {
+        variables: Map<String, Value>,
+        paths: Vec<Path>,
+    }
+
+    impl Variables {
+        fn new<'a>(
+            requires: Option<&Vec<Selection>>,
+            variable_usages: &[String],
+            data: &Value,
+            current_dir: &'a Path,
+            request: &Arc<Request>,
+            schema: &Arc<Schema>,
+        ) -> Result<Variables, FetchError> {
+            if let Some(requires) = requires {
+                let mut variables = Object::with_capacity(1 + variable_usages.len());
+                variables.extend(variable_usages.iter().filter_map(|key| {
+                    request.variables.as_ref().map(|v| {
+                        v.get(key)
+                            .map(|value| (key.clone(), value.clone()))
+                            .unwrap_or_default()
+                    })
+                }));
```

Comment:
:yarn: 
```suggestion
                let variables = variable_usages
                    .iter()
                    .filter_map(|key| {
                        request.variables.as_ref().map(|v| {
                            v.get(key)
                                .map(|value| (key.clone(), value.clone()))
                                .unwrap_or_default()
                        })
                    }))
                    .collect::<Object>();
```

Reply:
I tried that while doing the PR, but apparently not using with_capacity will have a performance impact (because it could trigger a reallocation when adding the last element)

Reply:
Oh I missed the `+ 1` sorry. You're :100: